appcypher-syntax.md

```diff
@@ -0,0 +1,947 @@
+    !=
+    not
+    and
+    or
+    in
+    not in
+    >>
+    <<
+    |
+    &
+    ~
+    ```
+
+
+- Custom operators are not supported because they are [easily abused](https://blog.jooq.org/2014/02/10/why-everyone-hates-operator-overloading/).
+
+
+## UNIFORM FUNCTION CALL SYNTAX
+- UNIFORM FUNCTION CALL SYNTAX (UFCS) is a feature that blurs the line between instance methods and regular functions.
+
+- A regular function can be called using both the dot notation call syntax or the regular function call syntax.
+
+    ```swift
+    func get_name(p: Person) -> String {
+        p.name
+    }
+
+    let person = Person::new("John", 52)
+
+    person.get_name() // Dot notation syntax
+    get_name(person) // Regular function call syntax
+    ```
+
+- Likewise instance methods can be called using both the dot notation call syntax or the regular function call syntax.
+
+    ```swift
+    impl Person {
+        func get_name(self) -> String {
+            self.name
+        }
+    }
+
+    let person = Person::new("John", 52)
+
+    person.get_name() // Dot notation syntax
+    get_name(person) //  Regular function call syntax
+    ```
+
+#### Why UFCS is useful
+
+Sometimes it makes sense to have certain functions be instance methods because they associate closely to a particular type. Other times it makes sense for such functions to be regular top-level functions because they don't associate with any single type. With UFCS, you don't have to memorize which one is a function or which one is an instance method, because you can call them using the both dot notation call and regular call syntaxes.
+
+
+- Take `map` function for example. It is a generic function and does not associate with any single type.
+
+    ```swift
+    func map[T, U](iter: T, fn: (U) -> U) -> T {
+        var new = T::new()
+        for (i, x) in enumerate(iter) {
+            new[i] = fn(x)
+        }
+        new
+    }
+    ```
+
+    Because we have UFCS support, we can chain up calls to regular function using the dot notation syntax, which can be easier to read sometimes.
+
+    ```swift
+    map(filter(map(arr, (x) => x * 2), (x) => x > 2), (x) => x * 2)
+    ```
+
+    ```swift
+    arr
+        .map((x) => x * 2)
+        .filter((x) => x > 2)
+        .map((x) => x * 2)
+    ```
+
+
+
+## TRAITS
+- Ratio does not support inheritance. Instead it choses composition-based type extension.
+
+    ```rust
+    trait Callable {
+        call: (...Int) -> Int
+    }
+    ```
+
+    ```swift
+    impl Sum: Callable {
+        func call(args: ...Int) -> Int {
+            var result = 0
+            for i in args {
+                result += i
+            }
+            result
+        }
+    }
+    ```
+
+## UNSAFE BLOCK
+- The unsafe block is necessary for scoping the usage of unsafe/unstable features like pointer arithmetic and destructive casting.
+
+    ```rust
+    let base_ptr = some_ref as *Int
+
+    let base_ptr = unsafe {
+        base_ptr.add(20)
+    }
+    ```
```

Comment:
Go, Rust, Swift, Java all have pointers. Rust and Go exposes them in a safe way. That is no pointer arithmetic, which may cause pointer aliasing problems. For unsafe stuff like pointer arithmetic, a lot of languages just lets you shoot yourself in the leg, but Rust does the right thing here with unsafe blocks. It limits the problem surface area, but you can still shoot yourself in the foot.

Unfortunately, when you need to go low-level, or interface C libraries, you need to have pointers in the language! If this is going to be a statically-typed language, pointers need to provided otherwise you won't be able to do a lot of low-level shenanigans.